File: protocol-generator/typescript/src/model/resource-target.ts
import { type EcfValue } from "../codec/ecf-value.js";
import { EntityProtocolError } from "../errors.js";
import * as Ecf from "./ecf.js";

/**
 * A `system/protocol/resource-target` (V7 §3.2): the data paths an operation
 * accesses (`targets`, at least one) plus optional `exclude` paths. The
 * dispatcher checks this against `grant.resources` before handler dispatch.
 */
export class ResourceTarget {
  constructor(
    readonly targets: readonly string[],
    readonly exclude: readonly string[] | null,
  ) {}

  toEcf(): EcfValue {
    return Ecf.map(
      ["targets", Ecf.array(this.targets.map((t) => Ecf.text(t)))],
      ["exclude", this.exclude === null ? null : Ecf.array(this.exclude.map((e) => Ecf.text(e)))],
    );
  }

  static fromEcf(value: EcfValue): ResourceTarget {
    const targets = Ecf.asArray(Ecf.require(value, "targets")).map((t) => Ecf.asText(t));
    if (targets.length === 0) {
      throw new EntityProtocolError("resource-target.targets MUST contain at least one entry (§3.2)");
    }
    const excludeField = Ecf.field(value, "exclude");
    const exclude = excludeField === null ? null : Ecf.asArray(excludeField).map((e) => Ecf.asText(e));
    return new ResourceTarget(targets, exclude);
  }
}

File: protocol-generator/csharp/src/EntityCore.Protocol/Model/ResourceTarget.cs
using EntityCore.Protocol.Codec;

namespace EntityCore.Protocol.Model;

/// <summary>
/// A <c>system/protocol/resource-target</c> (V7 §3.2): the data paths an operation
/// accesses (<c>targets</c>, at least one) plus optional <c>exclude</c> paths. The
/// dispatcher checks this against <c>grant.resources</c> before handler dispatch.
/// </summary>
internal sealed record ResourceTarget(IReadOnlyList<string> Targets, IReadOnlyList<string>? Exclude)
{
    public EcfValue ToEcf() => Ecf.Map(
        ("targets", Ecf.Array(Targets.Select(Ecf.Text))),
        ("exclude", Exclude is null ? null : Ecf.Array(Exclude.Select(Ecf.Text))));

    public static ResourceTarget FromEcf(EcfValue value)
    {
        IReadOnlyList<string> targets = Ecf.AsArray(Ecf.Require(value, "targets")).Select(Ecf.AsText).ToList();
        if (targets.Count == 0)
        {
            throw new EntityProtocolException("resource-target.targets MUST contain at least one entry (§3.2)");
        }
        EcfValue? exclude = Ecf.Field(value, "exclude");
        IReadOnlyList<string>? excludes = exclude is null
            ? null
            : Ecf.AsArray(exclude).Select(Ecf.AsText).ToList();
        return new ResourceTarget(targets, excludes);
    }
}
